### PR TITLE
switch keychain Set() to update existing items instead of delete/add

### PR DESCRIPTION
**Pull request by joemiller**
This switches the macos keychain's `Set()` implementation to use the go-keychain `UpdateItem()` API instead of doing a delete/re-add of the item.

Practically speaking this does not have much current impact to the library itself because it is not possible to retrieve this metadata through the current API, however, it sets the stage for improved metadata visibility like that being worked on in #29

example:
```
key                                      Created             LastModified
---------------------------------------- ------------------- -------------------
foo                                      2019-05-29 07:09:47 2019-05-29 07:27:21
```

### CODE REVIEW

File: keychain.go [hunk 1]
```diff
@@ -128,22 +129,32 @@ func (k *keychain) Set(item Item) error {
...
+		if err != nil {
+			return fmt.Errorf("Failed to query keychain: %v", err)
+		}
+		if len(results) == 0 {
+			return errors.New("no results")
 		}
 
-		debugf("Adding item again")
-		return gokeychain.AddItem(kcItem)
+		// Don't call SetAccess() as this will cause multiple prompts on update, even when we are not updating the AccessList
```
Comment (joemiller) [inline keychain.go]: We nil out the ACL here before calling update. If we don't do this, macOS will prompt a number of times and say incorrectly that we are changing ownership and changing ACLs even if the access list is identical to the currently stored item.

**Comment by lox** [comment]
This seems like a super sensible idea! I wonder why we didn't do this initially? 🤔 I dimly recall there was a reason we deleted and re-added, I'll see if I can figure out what it was and if it's still relevant.

**Comment by joemiller** [comment]
@lox It might be due to the inclusion of an accesslist in the UpdateItem call. It seems to cause macos-keychain to throw several prompts with text such as "the <app> wants to change ownership of an item" and "the <app> wants to change permissions of an item". This occurs even if the accesslist is identical to previous calls to Set(). To avoid this we need to call UpdateItem without an accesslist attached.

So with the code as-is, I think there is a tradeoff. You wouldn't be able to change the ACL on subsequent Set() calls after an item is initially created. You would need to remove and re-add in that case. I had a look at some of apple's docs and the keybase/go-keychain lib and I don't think there is a way to query for the ACL implemented in the go-keychain lib right now. It does not look like we get ACLs returned with QueryItem(). I think apple has separate calls to query item ACLs.

Still thinking about this a bit, but some ideas so far:

1. look into extending keybase/keychain lib to support reading ACLs from Items, then update Set() api in this lib to first check the item's existing ACL before deciding whether to call UpdateItem() with or without an ACL attached.
2. Extend the API in this lib to include additional funcs.. maybe Update() that only updates content but not ACL. That might be confusing with Set() though.

**Comment by lox** [comment]
> So with the code as-is, I think there is a tradeoff. You wouldn't be able to change the ACL on subsequent Set() calls after an item is initially created. 

Have you tried this change with `aws-vault` to see if it has any impact?

**Comment by joemiller** [comment]
I have not, as I'm not an aws-vault user. I actually work mostly with
google cloud.

**Comment by lox** [comment]
Ah, gotcha! I'll check it out then. Apologies, I assumed. I forget there are a lots of folks that use this dependency outside of `aws-vault`. That was after all why we decided to extract it 😅

**Comment by joemiller** [comment]
And thank you for extracting it =) I was halfway down the path of writing
my own almost identical lib before stumbling upon this.

**Comment by mtibben** [comment]
I've tested this with aws-vault and setting is working as expected with new and existing keychain items

**Pull request merged**